Add QtFeedback plugin

Tomin1 opened the pull request:
Implement libngf based QtFeedback plugin that supports playing, pausing
and resuming of effects although it depends on the backend whether those
will work expectedly. It does not support changing intensity level
although ngfd might set different intensities for different effects.
There needs to be a mapping configuration of the effect names to actual
haptic effects on ngfd side.

See also https://github.com/sailfishos/ngfd/pull/2.

Tomin1 committed:
Update urls

Signed-off-by: Tomi Leppänen <tomi.leppanen@jolla.com>

Tomin1 committed:
[libngf] Add QtFeedback plugin. JB#55261

Implement libngf based QtFeedback plugin that supports playing, pausing
and resuming of effects although it depends on the backend whether those
will work expectedly. It does not support changing intensity level
although ngfd might set different intensities for different effects.
There needs to be a mapping configuration of the effect names to actual
haptic effects on ngfd side.

Signed-off-by: Tomi Leppänen <tomi.leppanen@jolla.com>

pvuorela commented on rpm/libngf-qt5.spec:
```diff
@@ -12,24 +12,35 @@ BuildRequires:  pkgconfig(Qt5Core)
 BuildRequires:  pkgconfig(Qt5DBus)
 BuildRequires:  pkgconfig(Qt5Qml)
 BuildRequires:  pkgconfig(Qt5Test)
+BuildRequires:  pkgconfig(Qt0Feedback)
```
Looks like there's both 0 and 5 available. 5 could be better?

Tomin1 replied:
Changed.

pvuorela commented on rpm/libngf-qt5.spec:
```diff
@@ -65,11 +76,20 @@ sed 's/Nemo.Ngf/org.nemomobile.ngf/' < declarative/qmldir > %{buildroot}%{_libdi
 
+%post qtfeedback -p /sbin/ldconfig
+
+%postun qtfeedback -p /sbin/ldconfig
```
No libraries in /usr/lib so these shouldn't be needed.

Tomin1 replied:
Removed these.

pvuorela commented on feedback/ngffeedback.h:
```diff
@@ -0,0 +1,75 @@
+#include <qfeedbackplugininterfaces.h>
+#include "ngfclient.h"
+
+Q_DECLARE_LOGGING_CATEGORY(ngflc)
```
Doesn't matter too much but since it's only the single c++ file doing logging, I'd keep the logging details on that side until there's need elsewhere.

Tomin1 replied:
Yeah, removed this line.

pvuorela commented:
Looking at the code I only had some minor comments.

Tomin1 commented:
> Looking at the code I only had some minor comments.

I addressed your comments and made more changes based on what I discussed with @jusa. Now this can send more than one effect to ngfd and it will handle it. This is perhaps now a bit more complex than strictly required.

There are now to TODO comments on ngffeedback.cpp, please check those out if you have something to say about them.

Tomin1 commented on feedback/ngffeedback.cpp:
```diff
@@ -0,0 +1,318 @@
+        m_actuatorEnabled = value.toBool();
+        if (old != m_actuatorEnabled && !m_actuatorEnabled) {
+            /* Stop all effects */
+            for (auto it = m_activeEffects.begin(); it != m_activeEffects.end(); ++it) { // TODO: Would it make sense to iterate backwards?
```
So, what I mean here is that if ngfd was queueing these, would it be better to start dropping events from the end of the queue so that it could perhaps start playing the next event but it would not have to stop and continue to next all the time.

Tomin1 replied:
Ehm... There is one problem with my thinking: QHash is iterated in non-deterministic order.

Tomin1 replied:
Changed the underlying implementation to use QVector. Worse time-complexity but assuming a small number of items it shouldn't make a difference (why would you play a dozen effects at the same time?).

Tomin1 commented on feedback/ngffeedback.cpp:
```diff
@@ -0,0 +1,318 @@
+         */
+        quint32 id = m_customEffects.take(effect);
+        if (id) { // Existing effect
+            m_client.stop(id); // TODO: Does this make any sense?
```
Should I stop the effect so I can change the duration or can I just alter it somehow?

jusa replied:
Best to call stop for the event before starting a new one.

Tomin1 commented on feedback/ngffeedback.cpp:
```diff
@@ -0,0 +1,318 @@
+            for (auto it = m_activeEffects.begin(); it != m_activeEffects.end(); ++it) { // TODO: Would it make sense to iterate backwards?
+                quint32 id = it.key();
+                ActiveEffect effect = it.value();
+                it = m_activeEffects.erase(it); // Remove it already to avoid changing m_activeEffects
```
And need to fix this. I was thinking about it and then forgot to make sure it works correctly. The problem is that erase() returns iterator to next item, not previous so doing ++it will skip one. (iterating backwards with forward iterator would fix this in a quite simple way :wink:)

Tomin1 replied:
Yeah, well, ended up rewriting a good part of the plugin so maybe it works now.

jusa commented on feedback/ngffeedback.cpp:
```diff
@@ -0,0 +1,326 @@
+        if (!id)
+            qCWarning(ngflc) << "Could not play effect";
+        else
+            m_activeEffects.append({ id, QFeedbackEffect::Running, nullptr });
```
Like said I would probably not store effect state for these short effects, but it's fine this way as well.

I cancelled my previous pending review to start new one and apparently it also removed my comment I had added...

Tomin1 replied:
Yes, I understood that and it probably doesn't serve any purpose to store the status (since you can't actually retrieve it), but logging the events stopping might be useful still. If we don't care about logging that, then I can just drop this branch.

jusa replied:
Maybe log the id of started effect, then those can be matched by logging the id on effect states, even without having them in the array.

Tomin1 replied:
Right, I changed this and the signal handlers. I also added a comment about this.

jusa commented on feedback/ngffeedback.cpp:
```diff
@@ -0,0 +1,326 @@
+        return;
+
+    ActiveEffect *active = findCustomEffect(effect);
+    if (active)
```
Should this be if (!active) ? If not, add a comment :)

Tomin1 replied:
Eh... Yes, certainly. :) Will fix it.

jusa commented on feedback/ngffeedback.cpp:
```diff
@@ -0,0 +1,326 @@
+{
+    ActiveEffect *active = findEffect(id);
+    if (active) {
+        qCDebug(ngflc) << "Effect paused, id" << id;
```
I mean these could be outside the if, there shouldn't be a case where invalid id would be logged here.

Tomin1 replied:
I changed this so that it changes the state and then just logs the id. I had not realized that the ids from ngfd are app specific. It's nice that apps can't touch the effects from each other.
